File: java-algorithm-interview-with-kotlin/src/test/java/book/ch6/sub6/BookSolution.java
package book.ch6.sub6;

import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class BookSolution {

    @Test
    void test() {
        Solution solution = new Solution();
        String result = solution.longestPalindrome("dcbabcdd");
        Assertions.assertEquals(result, "dcbabcd");
    }


    class Solution {
        int left, maxLen;

        public String longestPalindrome(String s) {
            int len = s.length();
            if (len < 2) return s;

            for (int i = 0; i < len - 1; i++) {
                extendPalindrome(s, i, i + 1);
                extendPalindrome(s, i, i + 2);
            }
            return s.substring(left, left + maxLen);
        }

        private void extendPalindrome(String s, int j, int k) {
            while (j >= 0 && k < s.length() && s.charAt(j) == s.charAt(k)) {
                j--;
                k++;
            }

            if (maxLen < k - j - 1) {
                left = j + 1;
                maxLen = k - j - 1;
            }
        }
    }
}

File: java-algorithm-interview-with-kotlin/src/test/java/book/ch6/sub6/BookSolution.kt
package book.ch6.sub6

import org.junit.jupiter.api.Test

class BookSolution {

    class Solution {
        var left: Int = 0
        var maxLen: Int = 0

        fun extendPalindrome(s: String, j: Int, k: Int) {
            // 코틀린은 파라미터(value-parameter) 수정 불가
            var l = j
            var r = k

            while (l >= 0 && r < s.length && s[l] == s[r]) {
                l--
                r++
            }

            if (maxLen < r - l - 1) {
                left = l + 1
                maxLen = r - l - 1
            }
        }

        fun longestPalindrome(s: String): String? {
            val len = s.length

            if (len < 2) return s

            for (i in 0 until len - 1) {
                extendPalindrome(s, i, i + 1)
                extendPalindrome(s, i, i + 2)
            }
            return s.substring(left, left + maxLen)
        }
    }
}